### ISSUE TITLE
`np.innerproduct(a, a)` should be detected and dispatched to the BLAS `syrk` routine

### ISSUE THREAD

**Issue by jakirkham** [numpy/numpy#6948]
Related: https://github.com/numpy/numpy/issues/6794
Related: https://github.com/numpy/numpy/pull/6932

This is basically the same story as ( https://github.com/numpy/numpy/issues/6794 ) except for `np.innerproduct` instead. As the work has already been done for `np.dot` ( https://github.com/numpy/numpy/pull/6932 ), it should be a trivial operation to pull these changes over to `np.innerproduct` and create some similar tests.

**Comment by njsmith** [comment]
`inner` has always confused me, but staring at the definition, it looks like `inner` is just
```
def inner(a, b):
    b = np.asarray(b)
    if b.ndim >= 2:
        # Swap last two axes
        b = b.rollaxis(b.ndim - 1, b.ndim - 2)
    return np.dot(a, b)
```
(I haven't checked all the corner cases, e.g. mixing 1d and 2d inputs, and might have just misread entirely, but I think this handles most of the weird things; inner even matches dot's weird outer-product behavior for >=3d inputs.)

Where I'm going with this is, if `inner` doesn't share implementation with `dot` already, then maybe the right solution is not to remember to keep copying `dot` improvements over to `inner`, but rather to refactor them so that there's only one implementation to start with.

**Comment by jakirkham** [comment]
Sorry, I misspoke, I meant `innerproduct`. I updated the title and description to reflect that. That being said, if refactoring is an option, that sounds like the better way to go.

**Comment by seberg** [comment]
+1 was about to suggest the same. We kept forgetting bug fixes for inner before, and transpose should be cheap enough. Maybe there are also other ways to refactor, but any merging of the two would be great.

**Comment by jakirkham** [comment]
So, I lifted this straight from the documentation, but it states the following is true `np.inner(a, b) = np.tensordot(a, b, axes=(-1,-1))` for all `a` and `b` that have dimensionality greater than `0`. [1] In the case of both or either having `0` dimensionality, it looks like it is scalar multiplication with whatever rank object, which is also straightforward. As `np.tensordot` is in Python and uses `np.dot`, maybe `np.inner` can just be rewritten to explicitly use these definition. To be explicit, this is what I am thinking.

    def inner(a, b):
        if a.ndim != 0 and b.ndim !=0:
            return np.tensordot(a, b, axes=(-1,-1))
        else:
            return a*b

* [1] - http://docs.scipy.org/doc/numpy-1.10.0/reference/generated/numpy.inner.html

**Comment by jakirkham** [comment]
In any event, if our plan is to rewrite a simplified `inner`, what are we going to do with `cblas_innerproduct`? Deprecate it? Is it part of the formal API? If not, can we just delete it?

**Comment by njsmith** [comment]
The only things in the formal api are python functions, plus c struct
layouts, enum and #define values, and c functions marked with a comment
saying NUMPY_API or UMATH_API. (There's an automated code generator that
scans for those comments and uses them to export stuff.) Everything else we
can rearrange however we like.

**Comment by jakirkham** [comment]
I guess the `NPY_NO_EXPORT` should have been a hint or is that unrelated? In any event, thanks for clarifying.

**Comment by mhvk** [comment]
+1 one passing `inner`, etc., through `dot` -- if one also is careful with using `asanyarray` that means that with the introduction of `__numpy_ufunc__` all those will become subclass safe.

**Comment by jakirkham** [comment]
So, `inner` was originally in `numpy.core.multiarray`, which is a shared object. If we move it to Python, I don't think we can put it in this shared object any more or at least not easily, no? My vote would be rename `multiarray.so` to `_multiarray.so` and create a shim Python file `multiarray.py` that imports all of the contents from `_multiarray.so` and includes the Python definition of `inner`. Something similar may need to be done for `multiarray_tests.so`. Does this sound ok or does somebody have a better way of doing this?

**Comment by pv** [comment]
Tensordot's python argument processing probably has several ms overhead, so it can be slow compared to inner for small arrays.

**Comment by jakirkham** [comment]
It is slower, but I'm not sure about a `ms`. I am seeing maybe 1 order of magnitude slow down when trying a bunch of small arrays. For instance, `numpy.inner` is around ~1us; whereas, the implementation I placed above is taking ~10us. That being said, 1 order of slow down magnitude is not great.

Alternatively, @njsmith implementation sees a slowdown, of maybe a factor of 3 or 4 in the worse cases, but needs some help with a few of the corner cases. So, we aren't ready to really profile this yet and handling these corner cases may come at a penalty.

This still seems less than ideal as we are getting a slowdown while trying to optimize a special case. I would like to repropose the simple solution of copying the few lines of code from `cblas_innerproduct` to `cblas_matrixproduct`. If this really makes people cringe, that few lines of code can be refactored into an internal helper function that is used by both. The larger issue of refactoring `inner` to use `dot` is, in my mind, somewhat orthogonal to using this optimization.

**Comment by jakirkham** [comment]
To try one last Python implementation, this shakes off most of the unneeded `tensordot` stuff and reduces it down to the basic required operations. Now, to compare to the current implementation. For really small arrays, this ends up being 4 times slower. For smallish arrays, it ends up being ~1.5 times slower.

    def inner(a, b):
        a = np.asanyarray(a)
        b = np.asanyarray(b)
        if a.ndim > 0 and b.ndim > 0:
            at = a.reshape(-1, a.shape[-1])
            bt = np.rollaxis(b, -1, 0).reshape(b.shape[-1], -1)
            return np.dot(at, bt).reshape(a.shape[:-1] + b.shape[:-1])[()]
        else:
            return a*b

**Comment by njsmith** [comment]
@jakirkham: nobody should be referring to multiarray directly. It proba

**Comment by njsmith** [comment]
Sorry, fat fingered.

**Comment by njsmith** [comment]
@jakirkham: nobody should be referring to multiarray directly. It probably should have an underscored name, but with a few exceptions like linalg and testing then I think most internal module names are considered internal. (I guess the acid test is whether we get bug reports when we try changing stuff, but this is what I'd assume until then.) The idea of going through and underscoring some of these names is interesting actually...

NPY_NO_EXPORT is actually orthogonal -- that just prevents things from being exported using the standard system linker and should be applied to everything. (The functions that we export in the C API use a custom method that doesn't involve the system linker, for reasons that would take us pretty far off topic.)

@pv: Do you think we really have to worry about small constant overheads for these functions? There are substantial advantages to consolidating, both in terms of maintenance and (as @mhvk notes) semantics.

**Comment by jakirkham** [comment]
Thanks for clarifying the `multiarray` question and `NPY_NO_EXPORT`, @njsmith. So, where do you think `inner` should go (assuming you are still in favor of a Python solution instead of the C code)?

> @pv: Do you think we really have to worry about small constant overheads for these functions? There are substantial advantages to consolidating, both in terms of maintenance and (as @mhvk notes) semantics.

I'd like to second this point on maintenance and explicitly add readability. When I looked at `cblas_innerproduct` to introduce this optimization, I was eager to put it off because the code is quite long and not totally straightforward. Whereas, the 9 lines of Python above isn't bad to read and after written and tested, it has given me a better insight into what `inner`. While an order of magnitude loss does not seem acceptable (to me), a small loss in speed ~1.5 seems tolerable for these added advantages.

**Comment by pv** [comment]
@njsmith: I don't know in this specific case, just thought to mention the us+ performance overhead so that it gets considered. I suspect `np.inner` does not matter that much, as I'd guess the equivalent operation is more often written via `*` or `dot`.

**Comment by seberg** [comment]
Well, frankly the function is so simple.... Just do the transpose in C, it really is not a lot of code. Then you don't have to worry about interface changes, etc. Possible it is even exposed on the C-level, calling back into python is going to be more difficult then that.

**Comment by jakirkham** [comment]
Honestly, @seberg that seems like a fair suggestion (at least if I am understanding it correctly). My understanding is that you would like `cblas_innerproduct` to be refactored to use `cblas_matrixproduct`. I have tried to do this refactoring in this PR ( https://github.com/numpy/numpy/pull/6968 ).

**Comment by jakirkham** [comment]
So, sorry I went added this PR, but it seems like we were still discussing. After @seberg made his case, in particular, regarding `inner` possibly being exposed in the C API, I went and took a look and that does appear to be the case. ( http://docs.scipy.org/doc/numpy-1.10.0/reference/c-api.array.html#c.PyArray_InnerProduct )

Unless we want to have some Python code called from C, it seems like we should make the changes in C. That being said, if anyone has a different idea about how this should be implemented so as to make sure this C and Python API work, we should probably discuss this.
